packages/component-base/src/data-provider-controller/data-provider-controller.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * @license
+ * Copyright (c) 2021 - 2023 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { Cache } from './cache.js';
+import { getFlatIndexByPath, getFlatIndexContext } from './helpers.js';
+
+export class DataProviderController extends EventTarget {
+  constructor(host, { size, pageSize, isExpanded, dataProvider, dataProviderParams }) {
+    super();
+    this.host = host;
+    this.size = size;
+    this.pageSize = pageSize;
+    this.isExpanded = isExpanded;
+    this.dataProvider = dataProvider;
+    this.dataProviderParams = dataProviderParams;
+    this.rootCache = this.#createRootCache();
+  }
+
+  get effectiveSize() {
+    return this.rootCache.effectiveSize;
+  }
+
+  get #cacheContext() {
+    return { isExpanded: this.isExpanded };
+  }
+
+  isLoading() {
+    return this.rootCache.isLoading;
+  }
+
+  setSize(size) {
+    this.size = size;
+    this.rootCache.size = size;
```

Comment:
Unfurtunately, yes. The grid currently expects that `clearCache()` creates an empty cache with the `grid.size` property as the size, even if the size has changed since then as a result of a page request.

Reply:
This behavior can probably be considered a bug, but there are tests that rely on this behavior.